File: CSharp/leetcodesln/LinkedListCycle.cs
namespace leetcodesln
{
    public class LinkedListCycle
    {
        public bool HasCycle(ListNode head)
        {
            if (head == null) return false;

            var node_fast = head;
            var node_slow = head;

            while (node_fast.next != null && node_slow.next != null)
            {
                if (node_fast.next.next == null) return false;
                node_fast = node_fast.next.next;
                node_slow = node_slow.next;
                if (node_fast == node_slow) return true;
            }
            return false;
        }
    }
}

File: Go/questions/LinkedListCycle.go
package questions

import "github.com/leetcode/common"

func HasCycle(head *common.ListNode) bool {
	if head == nil {
		return false
	}

	fast, slow := head, head

	for fast.Next!=nil && slow.Next!=nil {
		if fast.Next.Next == nil {
			return false
		}

		fast = fast.Next.Next
		slow = slow.Next

		if fast == slow {
			return true
		}
	}
	return false
}
